sdk/bcs/src/index.ts

```diff
@@ -371,28 +396,134 @@ function ulebDecode(
+ * as well as an argument type for `bcs.registerStructType`.
+ */
+export type StructTypeDefinition = { [key: string]: string };
+
+/**
+ * Enum type definition. Used as input format in BcsConfig.types
+ * as well as an argument type for `bcs.registerEnumType`.
+ *
+ * Value can be either `string` when invariant has a type or `null`
+ * when invariant is empty.
+ *
+ * @example
+ * bcs.registerEnumType('Option<T>', {
+ *   some: 'T',
+ *   none: null
+ * });
+ */
+export type EnumTypeDefinition = { [key: string]: string | null };
+
+/**
+ * Configuration that is passed into BCS constructor.
+ */
+export type BcsConfig = {
+  /**
+   * Defines type name for the vector / array type.
+   * In Move: `vector<T>` or `vector`.
+   */
+  vectorType: string;
+  /**
+   * Address length. Varies depending on a platform and
+   * has to be specified for the `address` type.
+   */
+  addressLength: number;
+
+  /**
+   * Custom encoding for address. Supported values are
+   * either 'hex' or 'base64'.
+   */
+  addressEncoding?: 'hex' | 'base64';
+  /**
+   * Opening and closing symbol for type parameters. Can be
+   * any pair of symbols (eg `['(', ')']`); default value follows
+   * Rust and Move: `<` and `>`.
+   */
+  genericSeparators?: [string, string];
+  /**
+   * Type definitions for the BCS. This field allows spawning
+   * BCS instance from JSON or another prepared configuration.
+   * Optional.
+   */
+  types?: {
+    structs?: { [key: string]: StructTypeDefinition },
+    enums?: { [key: string]: EnumTypeDefinition }
+  }
+};
+
 /**
  * BCS implementation for Move types and few additional built-ins.
  */
-export class bcs {
+export class BCS {
   // Prefefined types constants
   static readonly U8: string = 'u8';
+  static readonly U16: string = 'u16';
   static readonly U32: string = 'u32';
   static readonly U64: string = 'u64';
   static readonly U128: string = 'u128';
+  static readonly U256: string = 'u256';
   static readonly BOOL: string = 'bool';
   static readonly VECTOR: string = 'vector';
   static readonly ADDRESS: string = 'address';
   static readonly STRING: string = 'string';
 
-  public static types: Map<string, TypeInterface> = new Map();
+  public types: Map<string, TypeInterface> = new Map();
+
+  /**
+   * Language-specific pair of separators used for defining generics.
+   * In Move and Rust this pair would be: `['<', '>']`.
+   */
+  public genericSeparators: [string, string];
+
+  // Set of values for forking BCS implementation.
+  protected vectorType: string;
+  protected addressLength: number;
+  protected addressEncoding: 'hex' | 'base64';
+
+  /**
+   * Construct a BCS instance with a prepared schema.
+   *
+   * @param schema A prepared schema with type definitions
+   * @param withPrimitives Whether to register primitive types by default
+   */
+  constructor(public schema: BcsConfig, withPrimitives: boolean = true) {
+    this.genericSeparators = schema.genericSeparators || ['<', '>'];
```

Comment:
If you want to make cloning easier, you could just stash the entire schema instead of assigning the properties into the instance.